### PR TITLE
[FIX] mail: handle transient messages with (un)seen notions

### PR DESCRIPTION
**Pull request by xdu-odoo**
- do not count transient messages in unread counter computation
- do not call channel_seen on transient messages

task-2343887

### CODE REVIEW

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js
Comment (xdu-odoo) [inline addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:669]: Found this quite verbal but didn't find another way to do that.
Not sure either if the component hint I used it the good one even if it does exactly what we want.

What do you guys think @seb-odoo & @alexkuhn ?

File: addons/mail/static/src/models/thread/thread.js
Comment (xdu-odoo) [inline addons/mail/static/src/models/thread/thread.js:1093]: This prevents flickering (1) in chat window

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js
Comment (seb-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:669]: This should probably be flux specific in the method calling the command, just like the other occurrence of `current-partner-just-posted-message` is in message post itself.

We don't want all tabs or even all viewers to sync with this. Only the specific thread viewer that was used to post the message should have it, as visual feedback the action was correctly done.

All other tabs or viewers can be left to the normal scrolling behavior on new message (scroll to bottom if the list was already scrolled to the bottom, which has to be fixed in task-2336251).

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:568]: Shouldn't we actually just convert id to `Math.floor(messageId)` at the start of the method, and let the rest of the code work normally?

- I don't think it's a good idea to leak the float id in `lastSeenByCurrentPartnerMessageId`.
- And we might need to actually call message seen with the last real id, even in this case?

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1093]: Maybe we want a computed field that is like `orderedMessages` but without `isTransient` messages on it? And then use that in all the places where that makes sense?

Comment (xdu-odoo) [reply addons/mail/static/src/models/thread/thread.js:1093]: Maybe but how would you name that field lol ? "realOrderedMessages" ?

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1093]: I would keep the "real" one with the existing name probably, and introduce a new one such as `orderedMessagesIncludingTransient`?

Honestly I just dislike the concept of fake records, and we are really starting to see why.

File: addons/mail/static/src/models/thread/thread.js
Comment (xdu-odoo) [reply addons/mail/static/src/models/thread/thread.js:568]: That's what was done and the task is saying 

> It's impossible to mark command message as read due to its fake ID, and the RPC to message_seen had to be fixed with Math.floor but it shouldn't be done instead.

So what is the purpose of this PR then ?

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:599]: In regard to this method, moving `floor` at the start instead of at the end?

Other fixes are still necessary too (auto scroll, transient message not considered when computing unread, ...)

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:1630]: You'll have to use this one when rendering messages in message list if we want transient messages to be shown.

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js
Comment (xdu-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:669]: I don't know how you can do that as the message is not posted at the moment you write the command. The message is created when you get the notification from the server. Should we create a "placeholder" message ?

Comment (seb-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:669]: `_handleNotificationPartnerTransientMessage` where we create the fake id would be a good place to put it then

Comment (seb-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:669]: resolving this one then

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js
Comment (seb-odoo) [inline addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:663]: maybe can iterate `message.threads` instead of doing this?

Comment (seb-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:663]: or just `nonOriginThreads` if we want to strictly match `channel_ids`

Comment (seb-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:663]: We don't support commands in non-channels?

Comment (seb-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:663]: Or `serverChannels` actually, what a mess all of those fields.

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js
Comment (seb-odoo) [inline addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:665]: I'm still not sure this is correct, maybe we should split the PR into what is obvious and what is not.

For the auto scroll, fixing it for normal messages as per task I linked earlier might be enough.

Comment (seb-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:665]: This is task-2336251, I think it's best if we block the current PR until that one is done. Things will become clearer after.

Comment (seb-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:665]: I'm taking it, I'll let you know when it's done and update this task spec accordingly.

File: addons/mail/static/src/models/thread/thread.js
Comment (xdu-odoo) [reply addons/mail/static/src/models/thread/thread.js:1630]: In fact message list is using `orderedMessages` from threadCache which contains transients

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1626]: Ok but this new field is not used anywhere then?

Comment (xdu-odoo) [reply addons/mail/static/src/models/thread/thread.js:1626]: Should thread_preview consider transient messages to determine last message ?
The same question for thread last message.

If they don't, then no, this field is not used and should not.
If they do, then no, this field is not used but should be.

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1626]: Thread preview should show the last real message as last message, not the transient message.

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1626]: And for the "thread last message" what do you mean?

Comment (xdu-odoo) [reply addons/mail/static/src/models/thread/thread.js:1626]: There is a computed field on thread model "lastMessage" which is based on orderedMessages which is also used in thread preview, thread needaction preview and notification list

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1626]: For all of those that shown on the messaging menu it's good to ignore transient, so I guess you can remove this new field here and just ensure the original one is excluding transient.

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js
Comment (xdu-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:666]: Is it really better ? I thought we could just trust the server here ?

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js
Comment (xdu-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:668]: As I understood, task-2336251 is waiting for https://www.odoo.com/web#cids=1&id=2341740&menu_id=&model=project.task&view_type=form

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js
Comment (seb-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:666]: The scroll part will be fully handled in task-2336251 you can remove it

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js
Comment (seb-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:668]: Yes, it's highlighting the bug you fixed of flickering new message so we can't merge it without the fix.

The current PR here (about commands) is not as much priority, focus on the task you just linked first.

Comment (xdu-odoo) [reply addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:668]: You should have a look as it's "ready" except a pending comment https://github.com/odoo/odoo/pull/57997

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js
Comment (seb-odoo) [inline addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:653]: `Math.max(lastMessageId, message.id)` ?

File: addons/mail/static/src/models/thread/thread.js
Comment (alexkuhn) [inline addons/mail/static/src/models/thread/thread.js:584]: Would clarify in parameter command that genuine messages have integer as id, and transient messages have decimals. Transient messages do not exist in server, which is why actual mark as seen on conversation for server must be applied to last genuine messages.

File: addons/mail/static/src/models/thread/thread.js
Comment (alexkuhn) [inline addons/mail/static/src/models/thread/thread.js:1625]: Why ignore transient messages here? This doesn't look right at a glance.

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/models/thread/thread.js
Comment (alexkuhn) [reply addons/mail/static/src/models/thread/thread.js:1625]: I guess it has to do with `orderedMessages` being used for unread and seen, and code breaks with transient messages.
Code that needs ordered messages without transient should have a field named appropriately.
I wouldn't mind having 2 fields `orderedMessages` and `orderedNonTransientMessages`.

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1625]: We had discussion before. We don't ever need the field with with the transient messages.

The only place where we use transient them is from thread cache to display them on message list.

Comment (alexkuhn) [reply addons/mail/static/src/models/thread/thread.js:1625]: `orderedMessages` should just be `messages` with some order, the fact it filters some messages mean that it should have a more specific name.

Comment (xdu-odoo) [reply addons/mail/static/src/models/thread/thread.js:1625]: so ?

Comment (alexkuhn) [reply addons/mail/static/src/models/thread/thread.js:1628]: Could rename this field to `orderedNonTransientMessages`.
I definitely don't like `orderedMessages` with no transient messages.

File: addons/mail/static/src/models/thread/thread.js
Comment (alexkuhn) [inline addons/mail/static/src/models/thread/thread.js:582]: format:
```js
/**
 * Mark the specified conversation as read/seen.
 *
 * @param {float} messageId the message to be considered as last seen.
 *   Genuine messages have integer id but transient ones have decimals.
 */
```

File: addons/mail/static/src/models/thread/thread.js
Comment (alexkuhn) [inline addons/mail/static/src/models/thread/thread.js:1526]: `lastMessage` is not actually last message, but last non transient message. Should also be renamed to be more specific (e.g. `lastNonTransientMessage`).

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1526]: Ok for renames if we do it quickly, but renaming fields in stable is not something we should do.

Comment (alexkuhn) [reply addons/mail/static/src/models/thread/thread.js:1526]: We don't have the same problems as with python fields. Rename is also a fix, because these fields were intentionally used just for genuine messages.

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1526]: We don't have DB problems but fields should be considered the same as public attributes / methods (which means no API change) to avoid breaking custom fixes/code based on them (either on our prod or by clients).

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1526]: I think the best when it's not an issue is just to create new fields. Here it wouldn't hurt to keep the unused field and add a new one.

Comment (alexkuhn) [reply addons/mail/static/src/models/thread/thread.js:1526]: > I think the best when it's not an issue is just to create new fields. Here it wouldn't hurt to keep the unused field and add a new one.

Yes: `orderedMessages` (keep old) + `orderedNonTransientMessages` (new field) + replace uses of `orderedMessages` by `orderedNonTransientMessages` when needed for fix. That's the step if we want to keep `orderedMessages` being as all ordered messages.

Previous discussion resulted in just renaming `orderedMessages` to `orderedNonTransientMessages`, so a change of meaning of a field may also imply a change of meaning of fields that rely on this, which is the case here with `lastMessage`.

Comment (seb-odoo) [reply addons/mail/static/src/models/thread/thread.js:1526]: The last message from @alexkuhn is a summary. Keep all existing fields. Add new fields for non transient. Change existing field usage to non transient fields in the code where it is appropriate. If computed fields (such as lastMessage) directly depend on a new transient field, create a new computed field for it too.

The rule is we need fields excluding transient almost everywhere, but we are in stable so we don't want to change the meaning of existing fields when possible, so we create new fields instead.

Comment (alexkuhn) [reply addons/mail/static/src/models/thread/thread.js:1526]: > Can one of you summarize what should be modified because I'm not sure I understood all of this discussion (or maybe just force push if little changes) so that we can move forward to the merge ?

My initial comment was about `orderedMessages` not matching its name with changes in PR. Decision was to rename to `orderedNonTransientMessages`. With this change, `lastMessage` field also doesn't match its name, and should be renamed to `lastNonTransientMessage`.

SEB highlights field rename must be discouraged in stable, so another way to fix the issue while not renaming fields is to keep old `orderedMessages` without filter on transient messages and introduce another computed field `orderedNonTransientMessages`.

**Review by seb-odoo — CHANGES_REQUESTED** [review]

**Comment by seb-odoo** [comment]
Also at least need 2 tests with commands, one for the counter and one for the new message separator.

**Comment by seb-odoo** [comment]
Considering the diff of the fix there should ideally be a test also for:

- thread preview last message & author prefix
- order in notification list
- author prefix on needaction preview

I would consider those less important but still nice to have since you actually fixed them too. Could also update commit text to include those.

**Comment by xdu-odoo** [comment]
Here are the two first tests, will work on the three others

**Comment by xdu-odoo** [comment]
Here are two more tests, I just don't find a way to test the needaction part (even if strongly identical to thread preview part). Could you give me a clue or explain more into details how needaction system works ?

**Comment by seb-odoo** [comment]
I'm not sure what issue you have with needaction but you can check that PR for fixed/updated needaction tests including mock server https://github.com/odoo/odoo/pull/59437/files

Note also task-2355918 because mock server for them is currently not ideal

File: addons/mail/static/src/components/thread_needaction_preview/thread_needaction_preview.xml
Comment (seb-odoo) [inline addons/mail/static/src/components/thread_needaction_preview/thread_needaction_preview.xml:44]: Wondering but shouldn't those actually be `lastNeedactionMessage` instead? cc @alexkuhn 

It's strange that in `o_ThreadNeedactionPreview_header` we use that field for the date, but then we display the last  message and not last needaction message which corresponded to the date.

Comment (alexkuhn) [reply addons/mail/static/src/components/thread_needaction_preview/thread_needaction_preview.xml:44]: Indeed, should be `lastNeedactionMessage` here.

Comment (xdu-odoo) [reply addons/mail/static/src/components/thread_needaction_preview/thread_needaction_preview.xml:44]: Then what would have to be tested in the scope of this PR for needaction's ?

Comment (alexkuhn) [reply addons/mail/static/src/components/thread_needaction_preview/thread_needaction_preview.xml:44]: Not sure to understand to question, but if it's how to reproduce the issue of not using `lastNeedactionMessage` here, it's when there's a needaction message that is not the last message of a conversation. You should see that author prefix uses author of last message instead of author of last needaction message. There's a mismatch with the inline body, which is on the last needaction message.

Comment (seb-odoo) [reply addons/mail/static/src/components/thread_needaction_preview/thread_needaction_preview.xml:44]: Test we see the proper last needaction message info everywhere, even if there is a more recent (not needaction) message.

Comment (seb-odoo) [reply addons/mail/static/src/components/thread_needaction_preview/thread_needaction_preview.xml:44]: Just like any other test, a hint that it covers the issue is that test should be red if you revert the change.

**Comment by xdu-odoo** [comment]
added test for needaction messages

**Review by alexkuhn — APPROVED** [review]
LGTM

File: addons/mail/static/src/models/thread/thread.js
Comment (seb-odoo) [inline addons/mail/static/src/models/thread/thread.js:1755]: missing dependency on orderedMessages.isTransient

File: addons/mail/static/tests/helpers/mock_server.js
Comment (seb-odoo) [inline addons/mail/static/tests/helpers/mock_server.js:851]: Isn't `ids` always `args.args[0]`?
And `commandName` is `args.kwargs.command || args.args[1]`

File: addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js
Comment (xdu-odoo) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: I think it's something @alexkuhn asked for

Comment (alexkuhn) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: Don't exactly remember what I said, but was wondering why comment below
`// need to remove focus from text area to avoid channel_seen`

And it's to reproduce the crash without fix. Thing is it's very technical to do this to make it crash without fix.
I don't like having tests that contain technical aspects like this.

Yes, in practice it crashes, and if we dig deeper the root cause is user interaction in tests are too simple (user click is not just a `.click()`, there's notably a focus change).

It's out of scope to improve user interaction simulations in tests. So just improving comment below is good enough for me. My mistake to let you have think I wanted to mark this test as "[technical]", which is not entirely true as I said above.

File: addons/mail/static/tests/helpers/mock_server.js
Comment (xdu-odoo) [reply addons/mail/static/tests/helpers/mock_server.js:851]: Does not seem to be something like that, once the command is passed as kwargs the other as args and once id is a list and the other one it's an int. Both seems to work.

image

File: addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js
Comment (seb-odoo) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: > Thing is it's very technical to do this to make it crash without fix.

It's not really technical though, it's our spec for mark as seen that depends on focus that comes from a functional decision.

File: addons/mail/static/tests/helpers/mock_server.js
Comment (seb-odoo) [reply addons/mail/static/tests/helpers/mock_server.js:851]: The one on the right is the server being smart and `browse` accepting more than the expected list of ids (int, newid or even string).

I would prefer if we changed it to always send a list of id, so we don't need to handle that in the mock server.

File: addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js
Comment (alexkuhn) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: Sorry, meant it looks technical. Rest of comment clearly tells me the root cause is simplistic simulation of click in tests.

Basically what I wanted to say with my comment is that comment is badly worded: only the person that made the test / investigated the bug initially will understand it. So the LOC looks out of place, therefore looks "technical". That’s the main problem I have, and even if that’s not "technical", the comment should be good enough to make this clear for everyone, or it feels as bad as this being technical without justification...

Comment (xdu-odoo) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: So just remove the [technical] or should we also change the comment ? I'm not sure what you're expecting in the comment then alex ?

Comment (alexkuhn) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: remove the `[technical]` tag in test name, and would change user interaction so that it's a user click, not a keydown "ENTER". As a result, the added `.focus()` would be much easier to get. Could still be worth keeping a comment above explicit `.focus()` change, briefly telling that an actual user click also changes the focus.

Comment (xdu-odoo) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: You cannot use click to send message in chat window ..

Comment (seb-odoo) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: There is no send button on the chat window, enter must be used. It's just that the user is clicking anywhere else after sending the message. It does not come from any technical reason, it comes from the fact it's the user interaction that we are simulating.

Comment (alexkuhn) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: So the name of test should reflect that... I mean, it's pretty obvious to me something is weird with test right now, the change of focus looks out of place. Please do the appropriate changes so that it doesn't look off, that the test is clearly intended with that. Could be name of test clearly mentioning message is posted while focus is somewhere else...

Comment (seb-odoo) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: Yes agree, this is my suggestion since the first message of this chain.

Comment (xdu-odoo) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: I could use some suggestion for test name and test comment I should add (and will make us win some time as I'm pretty sure my suggestions won't be as good as you wish :p )

Comment (seb-odoo) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: For the title I'd just add what I said originally: "unread counter in a chat window is not shown on receiving new transient message _even when composer is not focused_"

And then you can add a comment why we want to test that specific flux:
"The goal of removing the focus is to ensure the thread is not marked as seen automatically. Indeed that would remove the unread counter no matter what, which is already covered by other tests. The goal of this test is to cover the conditions specific to transient messages, and the conditions from focus would otherwise shadow them."

File: addons/mail/static/src/components/notification_list/notification_list.js
Comment (xdu-odoo) [reply addons/mail/static/src/components/notification_list/notification_list.js:107]: That's because we remove transient messages from the equation. Basically the separator show himself under the last message seen and as a transient message cannot be "seen" the separator can never be displayed somewhere else than under a message. So we need to add back these transient messages somehow in the equations. To ease all the process including server, I would suggest to have a boolean "lastMessageSeenByCurrentPartnerIsTransient" and a "lastSeenTransientMessage" and based on that show the separator at the right place.

The other solution is to accept that a transient message can be locally seen but prevent it from being told to server.
What approach would you prefer ?

File: addons/mail/static/src/models/thread/thread.js
Comment (xdu-odoo) [reply addons/mail/static/src/models/thread/thread.js:583]: it allows us to use short form in performRPC below

File: addons/mail/static/src/components/notification_list/notification_list.js
Comment (xdu-odoo) [reply addons/mail/static/src/components/notification_list/notification_list.js:107]: Added a skipped test that fails for now, need to make the decision to determine how to address this

File: addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js
Comment (seb-odoo) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js:1981]: Unresolving because I don't see the change. Did you push it?

File: addons/mail/static/src/components/notification_list/notification_list.js
Comment (seb-odoo) [reply addons/mail/static/src/components/notification_list/notification_list.js:107]: We probably need `lastSeenByCurrentPartnerMessageId` (same as now) and `lastSeenByCurrentPartnerNonTransientMessageId`, use the generic in the xml (same as before PR) and use the "non transient" one in the rest of the code (especially send to server).

Comment (seb-odoo) [reply addons/mail/static/src/components/notification_list/notification_list.js:107]: Best would be if separator was in model and we could decide where to place it depending on the following messages (we want to place it after the last seen message that is not a transient one).

Comment (seb-odoo) [reply addons/mail/static/src/components/notification_list/notification_list.js:107]: If this would be possible with your suggestion of `lastMessageSeenByCurrentPartnerIsTransient` feel free to try, I'm not sure what it implies, just do the one you think is best that works and we'll see from there.

File: addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js
Comment (xdu-odoo) [reply addons/mail/static/src/components/chat_window_manager/chat_window_manager_tests.js]: forgot the --force in my push, sorry ... I'll do the changes here and reopen on 14.0 later

File: addons/mail/static/src/components/notification_list/notification_list.js
Comment (xdu-odoo) [reply addons/mail/static/src/components/notification_list/notification_list.js:107]: Seems to work like this. If it seems ok I'll retarget

**Comment by seb-odoo** [comment]
Looks ok, make the 14.0 PR for the hopefully final check

**Comment by xdu-odoo** [comment]
retargeted to 14.0 https://github.com/odoo/odoo/pull/60083

**Pull request closed without merging**
